tests/unit/AutoEnforcer.cpp

```diff
@@ -0,0 +1,773 @@
+
+  BackingContainersStatus Targets;
+  Targets[CName].emplace_back("name", RootKind2);
+  auto Result = Step.cloneAndRun({});
+  BOOST_TEST(Result.get<MapContainer>(CName).get({ "name", RootKind2 }) == 1);
+}
+
+BOOST_AUTO_TEST_CASE(PipelineCanBeManuallyExectued) {
+  BackingContainerRegistry Registry;
+  Registry.addDefaultConstruibleFactory<MapContainer>(CName);
+
+  Pipeline Pip;
+  Pip.add(Step("first_step",
+               Registry.createEmpty(),
+               bindEnforcer<TestEnforcer>(CName, CName)));
+
+  Pip.getStartingContainer<MapContainer>(CName).get({ "name", RootKind }) = 1;
+
+  Pip.add(Step("End", Registry.createEmpty()));
+
+  BackingContainersStatus Targets;
+  Targets[CName].emplace_back("name", RootKind2);
+
+  auto Res = Pip[0].cloneAndRun({});
+  BOOST_TEST(Res.get<MapContainer>(CName).get({ "name", RootKind2 }) == 1);
+  Pip[0].mergeBackingContainers(std::move(Res));
+  const auto &StartingContainer = Pip.getStartingContainer<MapContainer>(CName);
+  auto Val = StartingContainer.get({ "name", RootKind2 });
+  BOOST_TEST(Val == 1);
+}
+
+BOOST_AUTO_TEST_CASE(SingleElementPipelineCanBeRunned) {
+  Pipeline Pip;
+
+  BackingContainers Containers;
+  Containers.add(CName, make_unique<MapContainer>());
+  Containers.get<MapContainer>(CName).get({ "name", RootKind }) = 1;
+
+  Pip.add(Step("first_step",
+               move(Containers),
+               bindEnforcer<TestEnforcer>(CName, CName)));
+  BackingContainers &BCI = Pip.back().getBackingContainers();
+  BOOST_TEST(BCI.get<MapContainer>(CName).get({ "name", RootKind }) == 1);
+
+  BackingContainers Containers2;
+  Containers2.add(CName, make_unique<MapContainer>());
+  Pip.add(Step("End", move(Containers2)));
+
+  BackingContainersStatus Targets;
+  Targets[CName].emplace_back("name", RootKind2);
+  auto Error = Pip.run(Targets);
+  BOOST_TEST(!Error);
+  BackingContainers &BC = Pip.back().getBackingContainers();
+  BOOST_TEST(BC.get<MapContainer>(CName).get({ "name", RootKind2 }) == 1);
+}
+
+class FineGranerEnforcer {
+
+public:
+  static constexpr auto Name = "FinedGranedEnforcer";
+  std::vector<InputOutputContract> getContract() const {
+    return {
+      InputOutputContract(RootKind, KindExactness::Exact, 0, FunctionKind, 1)
+    };
+  }
+
+  void run(const MapContainer &Source, MapContainer &Target) {
+    for (const auto &Element : Source.getMap()) {
+
+      if (Element.first.getKind() != RootKind)
+        continue;
+
+      auto Quantifiers = Element.first.getQuantifiers();
+      Quantifiers.emplace_back("f1");
+      Target.get({ move(Quantifiers), FunctionKind }) = Element.second;
+
+      Quantifiers = Element.first.getQuantifiers();
+      Quantifiers.emplace_back("f2");
+      Target.get({ move(Quantifiers), FunctionKind }) = Element.second;
+    }
+  }
+};
+
+BOOST_AUTO_TEST_CASE(SingleElementPipelineBackwardFinedGrained) {
+  PipelineRunner AE;
+  AE.addDefaultConstruibleFactory<MapContainer>(CName);
```

Comment:
As we discussed, can you rename this to something more along the lines of `registerDefaultConstructibleBackingContainerFactory`? This thing is not just a factory of backing containers, it's a factory for a *specific* (as in, with a specific name) backing container.